src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+}
+
+function processConstantModule(d) {
+	const moduleName = d[0][4].name;
+  const enclosedConstants = d[0][8];
+  // need to return array here to keep same signature as processComplexModule
+	return [{
+		name: moduleName,
+		definitions: enclosedConstants.flat(1),
+	}];
+}
+
+%}
+
+@lexer lexer
+
+main -> (importDcl __):* module:+ _ {% d => {
+	return d[1][0];
+}
+%}
+
+# support <import> or "import" includes - just ignored
+importDcl -> _ "#" "include" __ (%STRING | "<" _ %NAME ("/" %NAME):* "." "idl" _ ">") {% noop %}
+
+# constant modules need to be separate from complex modules since they shouldn't mix
+module  -> ((comment|annotation):* _ "module" __ fieldName __ "{" __ (constantDcl):+ __ "}" semi) {% processConstantModule %}
+  | ((comment|annotation):* _ "module" __ fieldName __ "{" __ (structWithAnnotation|module):+ __ "}" semi) {% processComplexModule %}
+
+
+structWithAnnotation -> (comment|annotation):* struct {% 
+ // ignore annotations on structs because we can only read default value (which doesn't apply)
+ d => {return d[1];}
+%}
+
+struct -> _ "struct" __ fieldName __ "{" __ (declaration):+ __ "}" semi {% d => {
+	const name = d[3].name;
+	const definitions = d[7].flat(2).filter(def => def !== null);
+	return {
+		name,
+		definitions,
+	};
+} %}
+
+constantDcl -> (comment|annotation):* constType semi {% d => d[1] %}
+declaration -> fieldWithAnnotation semi {% d => d[0] %}
+
+fieldWithAnnotation -> annotationOrCommentLines fieldDcl {% d=> {
+	let possibleAnnotations = [];
+	if(d[0]) {
+		possibleAnnotations = d[0];
+	}
+  const fields = d[1];
+	const finalDefs = fields.map((def) => extend([...possibleAnnotations, def]));
+	return finalDefs;
+} %}
+
+fieldDcl -> (
+     _ allTypes __  multiFieldNames arrayLength _
+   | _ allTypes __ multiFieldNames _
+   | _ sequenceType __ multiFieldNames _
+ ) {% (d) => {
+	const names = d[0].splice(3, 1)[0];
+	// create a definition for each name
+	const defs = names.map((nameObj) => extend([...d[0], nameObj]));
+	return defs
+} %}
+
+multiFieldNames -> fieldName (_ "," __ fieldName):* {%
+ d => {
+	 const fieldNames = d.flat(2).filter( d => d !== null && d.name);
+	 return fieldNames;
+ }
+%}
+   
+   
+   
+annotationOrCommentLines -> (annotation|comment):* {%
+  d => {
+	  return d[0][0] ? d[0][0].filter(d => d !== null) : null;
+  }
+%}
+
+annotation -> (
+    defaultAnnotation
+  | rangeAnnotation
+  | commentAnnotation
+  | keyAnnotation
+  | transferModeAnnotation
+  ) {% d => id(id(d)) %}
+
+defaultAnnotation -> _ at "default" _ "(" _ "value" assignment _ ")" {% d => { 
+	return {defaultValue: d[7].value };
+} %}
+
+# unsupported annotations are ignored
+rangeAnnotation -> _ at "range" _ "(" _ "min" assignment _ "," _ "max" assignment _ ")" {% noop %}
+
+commentAnnotation -> _ at "verbatim" _ "(" _ "language" assignment _ "," _ "text" _ %EQ _ STR ")" {% noop %}
+
+keyAnnotation -> _ at "key" {% noop %}
+
+transferModeAnnotation -> _ at "transfer_mode" _ "(" _ %NAME _ ")" {% noop %}
```

Comment:
surprised to see `_` at the beginning of all of these – maybe it should be added at a higher level? or is it here to make these rules more convenient to use elsewhere?